### ISSUE TITLE
UploadField hard limit of 30s timeout on uploads

### ISSUE THREAD

**Issue by HARVS1789UK** [silverstripe/silverstripe-asset-admin#977]
Seen in SilverStripe `4.4.2`

`UploadField` makes use of dropzonejs via `AssetDropzone`.

As per the [dropzonejs config documentation](https://www.dropzonejs.com/#config-timeout), dropzone implements a default timeout of 30 seconds for uploads.

There seems to be no way to change this default timeout through the `UploadField` class, and I am therefore left with a file upload field in the CMS which is not able to upload any files which take > 30 seconds to upload (in my case, videos of circa 100MB and up).

You can replicate this by setting up a fresh `4.4.2` install, adding a `has_many` -> `Files` to a DataObject and then configuring an `UploadField` to accept larger file sizes 
*Note* You will need to raise the default `php.ini`values for:
- `max_input_time`
- `max_execution_time`
- `max_post_size`
- `upload_max_filesize` 

to allow for larger files and longer times.
If your upload speed is so good it is hard to take > 30s to upload a file, I suggest using [Chrome's 'Network throttling` feature](https://css-tricks.com/throttling-the-network/) to simulate a slower connection.

The HTTP request to `upload` will be marked as `(cancelled)` after exactly 30s of processing the upload. This indicates to me that the client has cancelled the request.

The same behaviour can be seen in @kinglozzer's module `bigfork/silverstripe-dropzone` with the default dropzonejs config, but can be resolved by using his methods as below:

```
->setDropzoneConfig([
    'maxFilesize'   => (int)ini_get('upload_max_filesize'),
    'timeout'       => (int)ini_get('max_input_time') * 1000
])
```

It would be good to make these dropzonejs options configurable on `UploadField`, probably via the `AssetDropzone` JS  [referenced here](https://github.com/silverstripe/silverstripe-asset-admin/tree/1/client/src/components/AssetDropzone#options-object---required)

**Comment by adiwidjaja** [comment]
I had the same problem. I am leaving a comment with my server error message for anyone searching for the same error: Partial results are valid but processing is incomplete: [client xxx] mod_fcgid: can't get data from http client.

**Comment by HARVS1789UK** [comment]
@robbieaverill - Is there any roadmap for a fix for this?

I am using SS `4.2.2` for a project which I need to release in the next couple of weeks, so I am looking for a solution of some kind soon. That might end up being me forking asset-admin and putting in my own short term fix. If you could offer any guidance as to where I might put that patch it would be appreciated, as I couldn't find anything obvious when I was investigating the issue originally.

Not sure where to tie in to setting dropzone config options :-S

**Comment by kinglozzer** [comment]
@HARVS1789UK Have you tested what happens if you set the timeout to `0`, then rely on the server-side timeout to cancel (or error) the request? Dropzone doesn’t appear to add an `XMLHttpRequest.ontimeout` handler to display a timeout-specific error message, so if it always shows a generic error we could just hardcode the timeout to `0` rather than introduce new APIs for setting dropzone config values.

**Comment by HARVS1789UK** [comment]
@kinglozzer I can hardcode that change and give it a try, I suspect that might be in the return statement of getDefaultOptions() here? - https://github.com/silverstripe/silverstripe-asset-admin/blob/1/client/src/components/AssetDropzone/AssetDropzone.js

**Comment by HARVS1789UK** [comment]
@kinglozzer I am not confident that my suggestion above is the correct place to hardcode it, as looking through the object keys there vs the dropzonejs configuration options, they don't seem to tally up.

If it is/was the correct place, it's within `/client/src` anyway, which isn't present in my composer installed version in `/vendor` :-(

**Comment by kinglozzer** [comment]
@HARVS1789UK That does look like the correct place to me. If you install via composer, `client/src` won’t usually be included - though I _think_ it will be included if you use the `--prefer-source` flag.

**Comment by HARVS1789UK** [comment]
@kinglozzer I have just manually altered bundle.js  in `vendor/silverstripe/asset-admin/client/dist/js/bundle.js` to add `timeout:0,` in the relevant place.

This does seem to resolve my issue (I have tested by using Chrome's network throttling tool for over 3 mins then turning it off to allow the file to finish uploading and it doesn't timeout and completes successfully).

If I drop my PHP `post_max_size` value to 1MB (I am not sure which timing based setting I could change, as my `max_input_time` and `max_execution_time` settings are `60` and `30` respectively, yet that > 3min upload I just mentioned worked...) I get an error back as per the attached. (it is ill formatted, but I think that might be a separate SS issue?) 

<img width="1675" alt="Screenshot 2019-09-02 at 14 02 39" src="https://user-images.githubusercontent.com/5114645/64116745-0c2dff00-cd8b-11e9-92ae-6ad29b342a7b.png">

Could adding `timeout: 0,` to the default options here be a quick solution?

**Comment by kinglozzer** [comment]
Yeah that is a separate issue (#925)

What does the field currently look like if the request times out client-side? My only concern with this “hotfix” approach is that there’s no obvious way for the user to tell what went wrong if it _does_ time out server-side (I suspect they’ll see HTML like the above) 😕

**Comment by HARVS1789UK** [comment]
@kinglozzer Currently if the dropzonejs `timeout` config on the front-end is the culprit for the failed upload, there is no messaging whatsoever, the loading bar just stops moving and nothing is displayed to the end-user.

It is only once you inspect the underlying HTTP/XHR request you can see that it is `(cancelled)` as below:

<img width="1670" alt="Screenshot 2019-09-02 at 16 42 49" src="https://user-images.githubusercontent.com/5114645/64125147-f3304880-cda0-11e9-9c35-948935bb7847.png">

**Comment by kinglozzer** [comment]
Okay. So even if they see a broken error message with HTML, it’s obviously an error they can contact their developer about - unlike the status quo where they get no indication it’s failed - so I’m happy this will be an improvement. Fancy submitting a pull request against the 1.2 branch? I can try to help if you’ve not rebuilt from source before (some instructions are [here](https://docs.silverstripe.org/en/4/contributing/build_tooling/))

**Comment by ScopeyNZ** [comment]
Fixed by #985 . There's another issue that we identified that `UploadField` doesn't actually inherit use client config options in it's instance of `AssetDropzone`.
